### PR TITLE
feat(parachain/statement-distribution): Implement `receivedManifests`

### PR DESCRIPTION
**Pull request by haikoschol**
## Changes

This PR adds an implementation of the utility type `receivedManifests`, used in the grid tracker, which is part of the statement distribution subsystem.

## Tests

```sh
go test github.com/ChainSafe/gossamer/dot/parachain/statement-distribution
```

## Issues

#4574

### CODE REVIEW

**Review by freddyli7 — APPROVED** [review]
looks good to me

File: dot/parachain/statement-distribution/received_manifests.go [hunk 1]
```diff
@@ -0,0 +1,182 @@
...
+// The seconding limit is a per-validator limit. It ensures an upper bound on the total number of
+// candidates entering the system.
+//
+// The function mutates secondedCounts.
+func updatingEnsureWithinSecondingLimit(
+	secondedCounts map[parachaintypes.GroupIndex][]uint,
+	groupIndex parachaintypes.GroupIndex,
+	groupSize uint,
+	secondingLimit uint,
+	newSeconded parachaintypes.BitVec, // polkadot-sdk uses a &BitSlice<u8, Lsb0>
```
Comment (EclesioMeloJunior) [inline dot/parachain/statement-distribution/received_manifests.go]: I think the comment is not needed given our BitVec has the same func as the one from polkadot-sdk
```suggestion
	newSeconded parachaintypes.BitVec,
```

File: dot/parachain/statement-distribution/received_manifests.go [hunk 2]
```diff
@@ -0,0 +1,182 @@
...
+		secondedCounts[groupIndex] = counts
+	}
+
+	nsBits := newSeconded.Bits()
+	for i, bit := range nsBits {
+		if !bit {
+			continue
+		}
+
+		if i < len(counts)-1 && counts[i] == secondingLimit {
```
Comment (EclesioMeloJunior) [inline dot/parachain/statement-distribution/received_manifests.go]: `i < len(counts)` is enough given that when `i == len(counts)` it will be out of bounds
```suggestion
		if i < len(counts) && counts[i] == secondingLimit {
```

File: dot/parachain/statement-distribution/received_manifests.go [hunk 3]
```diff
@@ -0,0 +1,182 @@
...
+			return false
+		}
+	}
+
+	for i, bit := range nsBits {
+		if !bit {
+			continue
+		}
+
+		if i < len(counts)-1 {
```
Comment (EclesioMeloJunior) [inline dot/parachain/statement-distribution/received_manifests.go]:
```suggestion
		if i < len(counts) {
```

File: dot/parachain/statement-distribution/received_manifests.go [hunk 4]
```diff
@@ -0,0 +1,182 @@
...
+			counts[i] += 1
+		} else {
+			// polkadot-sdk does not contain this case and assumes groupSize == len(newSeconded)
+			// https://github.com/paritytech/polkadot-sdk/blob/3b4c48e7e3bba96091024643407994938607e3b9/polkadot/node/network/statement-distribution/src/v2/grid.rs#L865
```
Comment (EclesioMeloJunior) [inline dot/parachain/statement-distribution/received_manifests.go:172]: right, given that we are checking the seconding limit inside the group (to not allow more then X validators, from the same group, second the same candidate) so I believe that is why they don't assert the group len.

File: dot/parachain/types/bitvec.go [hunk 5]
```diff
@@ -249,3 +249,81 @@ func (bv *BitVec) Mask(mask BitVec) {
...
+		if i < len(otherBits) {
+			otherBit = otherBits[i]
+		}
+
+		// Set the result bit if either bit is set
+		resultBits[i] = thisBit || otherBit
+	}
+
+	return NewBitVec(resultBits)
+}
```
Comment (EclesioMeloJunior) [inline dot/parachain/types/bitvec.go]: Here is a suggestion to use the raw byte array and then perform the bitwise operation into a chunk of 8bits instead of transforming into a slice of bools and perform a per bit bitwise or. I've tested against your tests.
```suggestion
func (bv *BitVec) Or(other BitVec) BitVec {
	maxL := max(len(bv.bits), len(other.bits))
	result := make([]byte, maxL)
	copy(result, bv.bits[:])

	for i := 0; i < len(result); i++ {
		if i < len(other.bits) {
			result[i] |= other.bits[i]
		} else {
			break
		}
	}

	return BitVec {
		bits: result,
		len:  max(bv.len, other.len),
	}
}
```

Comment (haikoschol) [reply dot/parachain/types/bitvec.go]: Thanks! I had AI write it and then didn't even think of asking it to make it more efficient. I guess that's a case of AI brain rot. 🙈 
Also good call on not using the constructor, to get rid of the (in this case useless) `error` return value.

File: dot/parachain/statement-distribution/received_manifests.go (hunk 4 above)
Comment (haikoschol) [reply dot/parachain/statement-distribution/received_manifests.go:172]: What do you think, should we
* keep it as is
* replace `counts = append(counts, 1)` with a `logger.Warnf()`
* remove the check completely
?

File: dot/parachain/statement-distribution/received_manifests.go (hunk 2 above)
Comment (haikoschol) [reply dot/parachain/statement-distribution/received_manifests.go]: Good catch, thanks!

I've modified the test case `existing_counts_would_exceed_limit` to cover  this corner case.

File: dot/parachain/statement-distribution/received_manifests.go (hunk 4 above)
Comment (EclesioMeloJunior) [reply dot/parachain/statement-distribution/received_manifests.go:172]: I believe a logger.Warnf fits well!

**Review by EclesioMeloJunior — APPROVED** [review]
